// nevod/nevod-mongo-client.js
var MongoClient = require("mongodb").MongoClient,
    assert = require("assert");

var mongoURL = "mongodb://localhost:27017/uran"

module.exports.db = null

module.exports.connect = function(callback){
  if(this.db) return callback()
  MongoClient.connect(mongoURL,function(err,db){
    if(err) return callback(err)
    module.exports.db = db
    callback()
  })
}

module.exports.closeConnection = function(callback){
  if(this.db){
    this.db.close(function(err,result){
      this.db = null
      callback(err,result)
    })
  }
}

module.exports.writeDocsToDb = function(collection,docs,callback){
  if(docs.length==0 || docs == null || this.db == null){
    callback(false)
    return
  }
  this.db.collection(collection).insertMany(docs,function(err,result){
    callback(true,err,result)
  })
}

module.exports.findDocsInDb = function(collection,query,sorting,projection,callback){
  var cursor = this.db.collection(collection).find(query,projection).sort(sorting)
  var docs = []
  cursor.each(function(err,doc){
    if(doc!=null){
			docs.push(doc);
		} else {
			try {
				callback(docs)
        cursor.close()
			} catch (e) {
				console.log(e);
        cursor.close()
			}
		}
  })
}

module.exports.removeDocsFromDb = function(collection,query,callback){
  this.db.collection(collection).deleteMany(query,function(err,results){
    callback()
  })
}

module.exports.removeCollection = function(collection,callback){
  this.db.collection(collection).drop(function(err,reply){
    callback()
  })
}

module.exports.updateCollection = function(collection,query,update,upsert,callback){
  this.db.collection(collection).update(query,update,{upsert:upsert},function(){
    callback()
  })
}

module.exports.getDBCollections = function(options,callback){
  this.db.collections(function(error,cols){
    var names = []
    for(var i in cols){
      if(options.startsWith == "all"){
         names.push(cols[i].s.name)
      } else if(cols[i].s.name.startsWith(options.startsWith)){
         names.push(cols[i].s.name)
      }
    }
    callback(names)
  })
}
